src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -115,16 +113,182 @@ public CodeGenIL(IDictionary<string, IList> input, string filePath)
+            {
+                return EmitArrayCoercion<int, double>();
+            }
+            if (argType == typeof(double[]) && typeof(IEnumerable<long>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<double, long>();
+            }
+            if (argType == typeof(long[]) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<long, double>();
+            }
+            // TODO: The following cases do not work yet as EmitIListCoercion needs to be fixed.
+            //if (typeof(IList).IsAssignableFrom(argType) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<double>();
+            //}
+            //if (typeof(IList).IsAssignableFrom(argType)  && typeof(IEnumerable<int>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<int>();
+            //}
+            //if (typeof(IList).IsAssignableFrom(argType) && typeof(IEnumerable<long>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<long>();
+            //}
             // TODO: Add more coercion cases here.
 
             return argType;
         }
 
+        // TODO: Does not work yet (intended to coerce IList to IEnumerable<T>)
+        private Type EmitIListCoercion<T>()
+        {
+            MethodInfo mInfo = null;
+            // Find length for IList to be coerced (already on top of eval stack), len
+            var prop = typeof(ICollection).GetProperties(BindingFlags.Public | BindingFlags.Instance).Where(
+                                p => p.Name == nameof(ICollection.Count)).FirstOrDefault();
+
+            if (prop != null)
+            {
+                mInfo = prop.GetAccessors().FirstOrDefault();
+            }
+            //var mInfo = typeof(ICollection).GetMethod(nameof(ICollection.get_Count));
+            EmitOpCode(OpCodes.Callvirt, mInfo);
+
+            // var newarr = new T[len];
+            EmitOpCode(OpCodes.Newarr, typeof(T));
+
+            // Declare new array to store coerced values
+            var t = typeof(T).MakeArrayType();
+            var localBuilder = DeclareLocal(t);
+            var newArrIndex = localBuilder.LocalIndex;
+            EmitOpCode(OpCodes.Stloc, newArrIndex);
+
+            // Load array to be coerced.
+            EmitOpCode(OpCodes.Ldloc, localVarIndex);
+            // Emit new array
+            EmitOpCode(OpCodes.Ldloc, newArrIndex);
+
+            EmitOpCode(OpCodes.Ldc_I4_0);
+            mInfo = typeof(ICollection).GetMethod(nameof(ICollection.CopyTo));
+
+            // arr.CopyTo(newarr, 0);
+            EmitOpCode(OpCodes.Callvirt, mInfo);
+
+            EmitOpCode(OpCodes.Ldloc, newArrIndex);
+
+            return t;
+        }
+
+        private Type EmitArrayCoercion<S, T>()
```

Comment:
any reason not to have more descriptive type names?

Reply:
No reason, I just didn't think it was common to do that.